noria-server/src/controller/inner.rs

```diff
@@ -1146,60 +1142,64 @@ impl ControllerInner {
             leaf.index()
         );
 
-        if self
-            .ingredients
-            .neighbors_directed(leaf, petgraph::EdgeDirection::Outgoing)
-            .count()
-            > 0
-        {
-            // This query leaf node has children -- typically, these are readers, but they can also
-            // include egress nodes or other, dependent queries.
-            let mut has_non_reader_children = false;
-            let readers: Vec<_> = self
+        // We're looking for a single egress node that connects the query node to readers in
+        // other domains.
+        let mut nodes = vec![];
+        let egress_node = {
+            let num_children = self
                 .ingredients
                 .neighbors_directed(leaf, petgraph::EdgeDirection::Outgoing)
-                .filter(|ni| {
-                    if self.ingredients[*ni].is_reader() {
-                        true
-                    } else {
-                        has_non_reader_children = true;
-                        false
-                    }
-                })
-                .collect();
-            if has_non_reader_children {
-                // should never happen, since we remove nodes in reverse topological order
+                .count();
+            if num_children == 0 {
+                None
+            } else if num_children == 1 {
+                let child = self.ingredients
+                    .neighbors_directed(leaf, petgraph::EdgeDirection::Outgoing)
+                    .next()
+                    .unwrap();
+                assert!(self.ingredients[child].is_egress());
+                Some(child)
```

Comment:
This is the only branch that returns `Some()`. Why not just place the if statement from below right here instead?